src/docfx/restore/index/DependencySlotPool.cs

```diff
@@ -0,0 +1,211 @@
+                                    {
+                                        existed = true;
+                                        slot = i;
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (slot == null)
+                            {
+                                (acquired, acquirer) = await ProcessUtility.AcquireExclusiveLock(GetLockKey(url, $"{slots.Count + 1}"));
+                                if (acquired)
+                                {
+                                    slot = new T() { Id = $"{slots.Count + 1}" };
+                                }
+                            }
+
+                            Debug.Assert(slot != null && acquired && !string.IsNullOrEmpty(acquirer));
+
+                            // reset every property of rented slot
+                            slot.Url = url;
+                            slot.Restored = false;
+                            slot.LastAccessDate = DateTime.MinValue;
+                            slot.Acquirer = acquirer;
+
+                            slot = updateExistingSlot(slot);
+                            if (!existed)
+                                slots.Add(slot);
+                            break;
+                        case LockType.Shared: // find an matched slot for building
+                            foreach (var i in slots)
+                            {
+                                if (matchExistingSlot(i) && i.Restored)
+                                {
+                                    (acquired, acquirer) = await ProcessUtility.AcquireSharedLock(GetLockKey(url, i.Id));
+                                    if (acquired)
+                                    {
+                                        slot = i;
+                                        slot.Acquirer = acquirer;
+                                        break;
+                                    }
+                                }
+                            }
+                            break;
+                        default:
+                            throw new NotSupportedException($"{type} is not supported");
+                    }
+
+                    WriteSlots(restoreDir, slots);
+                });
+
+            if (slot != null)
+            {
+                Debug.Assert(!string.IsNullOrEmpty(slot.Acquirer));
+            }
+
+            return slot == null ? default : ($"{slot.Id}", slot);
+        }
+
+        public static List<T> GetSlots<T>(string restoreDir) where T : DependencySlot
+        {
+            Debug.Assert(!string.IsNullOrEmpty(restoreDir));
+
+            var slotsFile = Path.Combine(restoreDir, "index.json");
+            var content = File.Exists(slotsFile) ? File.ReadAllText(slotsFile) : string.Empty;
+
+            return JsonUtility.Deserialize<List<T>>(content) ?? new List<T>();
```

Comment:
consider use a wrapper object like other data models, it is hard to extend an array.

Reply:
ok

Reply:
ok